tests/test_config.py

```diff
@@ -0,0 +1,382 @@
+    )
+    def test_str_to_upper(self, enabled, str_check, result_str_check):
+        class Model(BaseModel):
+            str_check: str
+
+            class Config:
+                str_to_upper = enabled
+
+        m = Model(str_check=str_check)
+
+        assert m.str_check == result_str_check
+
+    @pytest.mark.parametrize(
+        'enabled,str_check,result_str_check',
+        [(True, 'ABCDefG', 'abcdefg'), (False, 'ABCDefG', 'ABCDefG')],
+    )
+    def test_str_to_lower(self, enabled, str_check, result_str_check):
+        class Model(BaseModel):
+            str_check: str
+
+            class Config:
+                str_to_lower = enabled
+
+        m = Model(str_check=str_check)
+
+        assert m.str_check == result_str_check
+
+    def test_namedtuple_arbitrary_type(self):
+        class CustomClass:
+            pass
+
+        class Tup(NamedTuple):
+            c: CustomClass
+
+        class Model(BaseModel):
+            x: Tup
+
+            class Config:
+                arbitrary_types_allowed = True
+
+        data = {'x': Tup(c=CustomClass())}
+        model = Model.model_validate(data)
+        assert isinstance(model.x.c, CustomClass)
+        with pytest.raises(PydanticSchemaGenerationError):
+
+            class ModelNoArbitraryTypes(BaseModel):
+                x: Tup
+
+    @pytest.mark.parametrize(
+        'use_construct, populate_by_name_config, arg_name, expectation',
+        [
+            [False, True, 'bar', does_not_raise()],
+            [False, True, 'bar_', does_not_raise()],
+            [False, False, 'bar', does_not_raise()],
+            [False, False, 'bar_', pytest.raises(ValueError)],
+            [True, True, 'bar', does_not_raise()],
+            [True, True, 'bar_', does_not_raise()],
+            [True, False, 'bar', does_not_raise()],
+            [True, False, 'bar_', does_not_raise()],
+        ],
+    )
+    def test_populate_by_name_config(
+        self,
+        use_construct: bool,
+        populate_by_name_config: bool,
+        arg_name: str,
+        expectation: ContextManager,
+    ):
+        expected_value: int = 7
+
+        class Foo(BaseModel):
+            bar_: int = Field(..., alias='bar')
+
+            class Config(BaseConfig):
+                populate_by_name = populate_by_name_config
+
+        with expectation:
+            if use_construct:
+                f = Foo.model_construct(**{arg_name: expected_value})
+            else:
+                f = Foo(**{arg_name: expected_value})
+            assert f.bar_ == expected_value
+
+    def test_immutable_copy_with_frozen(self):
+        class Model(BaseModel):
+            a: int
+            b: int
+
+            class Config:
+                frozen = True
+
+        m = Model(a=40, b=10)
+        assert m == m.copy()
+
+    def test_config_class_is_deprecated(self):
+        with pytest.warns(
+            DeprecationWarning, match='`BaseConfig` is deprecated and will be removed in a future version'
```

Comment:
Even though we didn't discuss it, I added a DeprecationWarning any time you subclass `BaseConfig` (since I figure there's a good chance people will do that without ever accessing the attributes). Let me know if you think we should drop it.

The next test shows the deprecation warnings that are raised for access of any valid config attributes, and the test after that shows the errors raised for attempting to access "invalid" attributes.

Reply:
LGTM.